source/common/upstream/health_discovery_service.cc

```diff
@@ -197,11 +199,35 @@ HostVectorConstSharedPtr HdsCluster::createHealthyHostList(const HostVector& hos
 ClusterInfoConstSharedPtr ProdClusterInfoFactory::createClusterInfo(
     Runtime::Loader& runtime, const envoy::api::v2::Cluster& cluster,
     const envoy::api::v2::core::BindConfig& bind_config, Stats::Store& stats,
-    Ssl::ContextManager& ssl_context_manager, Secret::SecretManager& secret_manager,
-    bool added_via_api) {
+    Ssl::ContextManager& ssl_context_manager, bool added_via_api, ClusterManager& cm,
+    const LocalInfo::LocalInfo& local_info, Event::Dispatcher& dispatcher,
+    Runtime::RandomGenerator& random) {
+
+  Envoy::Stats::ScopePtr scope = stats.createScope(fmt::format("cluster.{}.", cluster.name()));
+
+  Envoy::Server::Configuration::TransportSocketFactoryContextImpl factory_context(
+      ssl_context_manager, *scope, cm, local_info, dispatcher, random, stats);
+
+  auto transport_socket = cluster.transport_socket();
```

Comment:
This seems to be the meat of the change, I think it would be ideal to refactor this into a utility shared with other cluster implementations, since this seems shared code. Can you do this and also fix the merge conflicts? Thanks.